src/OpenTelemetry/Metrics/MeterProviderSdk.cs

```diff
@@ -247,33 +189,25 @@ internal sealed class MeterProviderSdk : MeterProvider
                             return;
                         }
 
-                        var meterName = instrument.Meter.Name;
-                        var metricName = instrument.Name;
-                        var metricStreamName = $"{meterName}.{metricName}";
-                        Metric metric = null;
-                        lock (this.instrumentCreationLock)
+                        if (this.reader != null)
                         {
-                            if (this.metricStreamNames.Contains(metricStreamName))
-                            {
-                                OpenTelemetrySdkEventSource.Log.MetricInstrumentIgnored(metricName, instrument.Meter.Name, "Metric name conflicting with existing name.", "Either change the name of the instrument or change name using View.");
-                                return;
-                            }
-
-                            var index = ++this.metricIndex;
-                            if (index >= MaxMetrics)
+                            if (this.compositeMetricReader == null)
                             {
-                                OpenTelemetrySdkEventSource.Log.MetricInstrumentIgnored(metricName, instrument.Meter.Name, "Maximum allowed Metrics for the provider exceeded.", "Use views to drop unused instruments. Or configure Provider to allow higher limit.");
-                                return;
+                                var metric = this.reader.AddMetricWithNoViews(instrument);
+                                if (metric != null)
+                                {
+                                    listener.EnableMeasurementEvents(instrument, metric);
+                                }
                             }
                             else
                             {
-                                metric = new Metric(instrument, temporality, metricName, instrument.Description);
-                                this.metrics[index] = metric;
-                                this.metricStreamNames.Add(metricStreamName);
+                                var metrics = this.compositeMetricReader.AddMetricsWithNoViews(instrument);
+                                if (metrics.Any(metric => metric != null))
+                                {
+                                    listener.EnableMeasurementEvents(instrument, metrics);
```

Comment:
we need to ensure that the MeasuementCompleted callback is expecting List<Metric> here.

Reply:
Fixed this.